tests/draft2019-09/ref.json

```diff
@@ -382,5 +382,70 @@
                 "valid": false
             }
         ]
+    },
+    {
+        "description": "order of evaluation: $id and $ref",
+        "schema": {
+            "$comment": "$id must be evaluated before $ref to get the proper $ref destination",
+            "$id": "/base/base.json",
+            "$ref": "int.json",
+            "$defs": {
+                "bigint": {
+                    "$comment": "canonical uri: /base/int.json",
+                    "$id": "int.json",
+                    "maximum": 10
+                },
+                "smallint": {
+                    "$comment": "canonical uri: /int.json",
+                    "$id": "/int.json",
+                    "maximum": 2
+                }
+            }
+        },
+        "tests": [
+            {
+                "description": "data is valid against first definition",
+                "data": 5,
+                "valid": true
+            },
+            {
+                "description": "data is invalid against first definition",
+                "data": 50,
+                "valid": false
+            }
+        ]
+    },
+    {
+        "description": "order of evaluation: $id and $anchor and $ref",
+        "schema": {
+            "$comment": "$id must be evaluated before $ref to get the proper $ref destination",
+            "$id": "/base.json",
+            "$ref": "#bigint",
+            "$defs": {
+                "bigint": {
+                    "$comment": "canonical uri: /base.json#/$defs/bigint; another valid uri for this location: /base.json#bigint",
```

Comment:
This test case perfectly covers one of the aspects of `$anchor` that I've been confused about. Thanks for this!

If `#/$defs/bigint` weren't here, would `/base.json#bigint` refer to `#/$defs/smallint`? Or would that fall under the category of undefined behavior where an implementation may raise an error? I can read the spec both ways. Specifically, this bit of [JSON Schema Core section 8.2.3](https://json-schema.org/draft/2019-09/json-schema-core.html#rfc.section.8.2.3):

>Two "$anchor" keywords **in the same schema document** MAY have the same value if they apply to different base URIs, as the resulting full URIs will be distinct. However, the effect of two "$anchor" keywords with the same value and the same base URI is undefined. Implementations MAY raise an error if such usage is detected. 

So it's undeniable that multiple anchors within the same schema document are allowed. But I had been thinking of "in the same schema document" as only being possible if the duplicate anchors were in different **resource** root documents. With that interpretation, this test case would be undefined behavior, and I could throw an error because two different schemas can be reached by the same anchor on the same base URI. Just like how I'd expect `/base.json#/$defs/smallint` to be equivalent to `/`, I'd expect `/base.json#bigint` to refer to both `#/$defs/bigint` and `#/$defs/smallint`.

Looking at the [previous section](https://json-schema.org/draft/2019-09/json-schema-core.html#embedded), though, it looks like implementations aren't actually required to recognize `/base.json#/$defs/smallint` as a reference to `/` here. This test's interpretation is consistent with that.

I'm honestly conflicted about which way I prefer, but I think both could be valid. I might implement a "strict" mode that enforces `$id` as a boundary, and one that allows JSON references to pass through `$id`s via a parent URI.

Does that make sense?